src/stages/stage5/spells/artificial_lightning.c

```diff
@@ -10,148 +10,155 @@
 
 #include "spells.h"
 
-MODERNIZE_THIS_FILE_AND_REMOVE_ME
+TASK(lightningball_particle, { BoxedBoss boss; }) {
+	Boss *boss = TASK_BIND(ARGS.boss);
+	for(int x = 0;; x++, YIELD) {
+		cmplx n = cdir(M_TAU * rng_real());
+		float l = 150 * rng_real() + 50;
+		float s = 4 + x * 0.01;
 
-static int zigzag_bullet(Projectile *p, int t) {
-	if(t < 0) {
-		return ACTION_ACK;
-	}
-
-	int l = 50;
-	p->pos = p->pos0+(abs(((2*t)%l)-l/2)*I+t)*2*p->args[0];
-
-	if(t%2 == 0) {
 		PARTICLE(
 			.sprite = "lightningball",
-			.pos = p->pos,
-			.color = RGBA(0.1, 0.1, 0.6, 0.0),
-			.timeout = 15,
-			.draw_rule = Fade,
+			.pos = boss->pos,
+			.color = RGBA(0.05, 0.05, 0.3, 0),
+			.draw_rule = pdraw_timeout_fade(0.5, 0),
+			.timeout = l/s,
+			.move = move_linear(-s * n),
 		);
 	}
+}
+
+TASK(zigzag_move, { BoxedProjectile p; cmplx move_arg; }) {
+	Projectile *p = TASK_BIND(ARGS.p);
+
+	for(int time = 0;; time++, YIELD) {
+		// TODO: make this better
+		p->pos = p->pos0 + (abs(((2 * time) % 50) - 50 / 2) * I + time) * 2 * ARGS.move_arg;
 
-	return ACTION_NONE;
+		if(time % 2 == 0) {
+			PARTICLE(
+				.sprite = "lightningball",
+				.pos = p->pos,
+				.color = RGBA(0.1, 0.1, 0.6, 0.0),
+				.timeout = 15,
+				.draw_rule = pdraw_timeout_fade(0.5, 0),
+			);
+		}
+	}
 }
 
-static int lightning_slave(Enemy *e, int t) {
-	if(t < 0)
-		return 1;
-	if(t > 200)
-		return ACTION_DESTROY;
+TASK(zigzag_shoot, { BoxedBoss boss; } ) {
+	Boss *boss = TASK_BIND(ARGS.boss);
+	for(int time = 0;; time++, YIELD) {
+		if(time > 0 && !(time % 100)) {
+			int count = difficulty_value(7, 7, 7, 9);
+			for(int i = 0; i < count; i++) {
+				Projectile *p = PROJECTILE(
+					.proto = pp_bigball,
+					.pos = boss->pos,
+					.color = RGBA(0.5, 0.1, 1.0, 0.0),
+				);
+				INVOKE_SUBTASK(zigzag_move, ENT_BOX(p), cdir(M_TAU / count * i) * cnormalize(global.plr.pos - boss->pos));
+			}
+			play_sfx("redirect");
+			play_sfx("shot_special1");
+		}
+	}
+}
+
+TASK(lightning_slave_move, { BoxedEnemy e; cmplx move_arg; }) {
+	Enemy *e = TASK_BIND(ARGS.e);
+	cmplx move = ARGS.move_arg;
 
-	TIMER(&t);
+	for(int time = 0;; time++, YIELD) {
+		e->move = move_linear(move);
+		if(time%20 == 0) {
+			move *= cdir(0.25 + 0.25 * rng_real() * M_PI);
+		}
+	}
+}
 
-	e->pos += e->args[0];
+TASK(lightning_slave, { cmplx pos; cmplx move_arg; }) {
+	// TODO: get rid of ENEMY_IMMUNE
+	Enemy *e = TASK_BIND(create_enemy1c(ARGS.pos, ENEMY_IMMUNE, NULL, NULL, 0));
 
-	FROM_TO(0,200,20)
-		e->args[0] *= cexp(I * (0.25 + 0.25 * frand() * M_PI));
+	INVOKE_TASK(lightning_slave_move, ENT_BOX(e), ARGS.move_arg);
 
-	FROM_TO(0, 200, 3)
-		if(cabs(e->pos-global.plr.pos) > 60) {
-			Color *clr = RGBA(1-1/(1+0.01*_i), 0.5-0.01*_i, 1, 0);
+	for(int x = 0; x < 67; x += WAIT(3)) {
```

Comment:
what is `x += WAIT(3)`? i am not a fan of these strange for-loop reinvention constructions. but if it has to be done, `x++, WAIT(3)` is better

Reply:
@laochailan that's not equivalent. `WAIT()` returns the amount of frames waited, so you'd have to do `x += 3, WAIT(3)` if you really wanted to rewrite this. But that's repeating yourself, and not any better in my opinion.

Reply:
then this is wrong because the amount of iterations has been shortened to 200/3.

Reply:
Ah yes, you are correct. I missed the context.

The `_i` that the `x` is replacing is the iteration count (i think), not a timer. So it makes no sense to add the wait time to it. It should indeed be `x++, WAIT(3)`, or more "conventionally":
```c
for(int x = 0; x < 67; x++) {
	/* the loop... */
	WAIT(3);
}
```